src/vs/workbench/contrib/snippets/browser/surroundWithSnippet.ts

```diff
@@ -4,57 +4,136 @@
+		return allSnippets.filter(snippet => snippet.usesSelection);
+	}
+
+	canExecute(): boolean {
+		return this._contextKeyService.contextMatchesRules(options.precondition);
+	}
 
-		if (!editor.hasModel()) {
+	async run() {
+		if (!this.canExecute()) {
 			return;
 		}
 
-		const { lineNumber, column } = editor.getPosition();
-		editor.getModel().tokenization.tokenizeIfCheap(lineNumber);
-		const languageId = editor.getModel().getLanguageIdAtPosition(lineNumber, column);
-
-		const allSnippets = await snippetService.getSnippets(languageId, { includeNoPrefixSnippets: true, includeDisabledSnippets: true });
-		const surroundSnippets = allSnippets.filter(snippet => snippet.usesSelection);
-		const snippet = await instaService.invokeFunction(pickSnippet, surroundSnippets);
+		const snippets = await this.getSurroundableSnippets();
+		if (!snippets.length) {
+			return;
+		}
 
+		const snippet = await this._instaService.invokeFunction(pickSnippet, snippets);
 		if (!snippet) {
 			return;
 		}
 
-
 		let clipboardText: string | undefined;
 		if (snippet.needsClipboard) {
-			clipboardText = await clipboardService.readText();
+			clipboardText = await this._clipboardService.readText();
 		}
 
-		SnippetController2.get(editor)?.insert(snippet.codeSnippet, { clipboardText });
+		SnippetController2.get(this._editor)?.insert(snippet.codeSnippet, { clipboardText });
+	}
+}
+
+class SurroundWithSnippetEditorAction extends EditorAction2 {
+	constructor() {
+		super(options);
+	}
+	async runEditorCommand(accessor: ServicesAccessor, editor: ICodeEditor, ...args: any[]) {
+		const instaService = accessor.get(IInstantiationService);
+		const core = instaService.createInstance(SurroundWithSnippet, editor);
+		await core.run();
+	}
+}
+
+registerAction2(SurroundWithSnippetEditorAction);
+
+
+export class SurroundWithSnippetCodeActionProvider extends Disposable implements CodeActionProvider {
+	private static readonly codeAction: CodeAction = {
+		kind: CodeActionKind.QuickFix.value,
+		title: options.title.value,
+		command: {
+			id: options.id,
+			title: options.title.value,
+		},
+	};
+
+	private core: SurroundWithSnippet;
+
+	constructor(
+		editor: ICodeEditor,
+		@ILanguageFeaturesService languageFeaturesService: ILanguageFeaturesService,
+		@IInstantiationService instaService: IInstantiationService,
+	) {
+		super();
+		this.core = instaService.createInstance(SurroundWithSnippet, editor);
+		this._register(languageFeaturesService.codeActionProvider.register('*', this));
 	}
-});
+
+	async provideCodeActions(model: ITextModel, range: Range | Selection, context: CodeActionContext, token: CancellationToken): Promise<CodeActionList> {
+		if (!this.core.canExecute()) {
+			return { actions: [], dispose: () => { } };
+		}
+		const snippets = await this.core.getSurroundableSnippets();
+		return {
+			actions: snippets.length ? [SurroundWithSnippetCodeActionProvider.codeAction] : [],
```

Comment:
This will just invoke the picker but I think it would be better to directly show the snippets. Only show the picker when there is too much to pick from (>7 or so). 

For each surroundable snippet you can use [InsertSnippetAction](https://github.com/microsoft/vscode/blob/1c5723822a33d7c40913f98d8a2f639b40e7f8ad/src/vs/workbench/contrib/snippets/browser/insertSnippet.ts#L48) and bind the snippet-argument accordingly